### ISSUE TITLE
Handle scalar with Krnl

### ISSUE THREAD

**Issue by chentong319** [onnx/onnx-mlir#872]
I discovered this issue when working on yolov5 model.  I got `error: Expected positive number of original loops.`
The model has some scalars with type `tensor<f32>` or `tensor<i64>`. Tense with scalar type is needed because some ops, for example, Clip, require scalar input. Moreover, some shape related ops, such as Squeeze/Unsqueeze, behave differently between tensor<i64> and tensor<1xi64>. Therefore, we can not convert scalar type to tensor with one element.
The issue with scalar type is how we access it. We build loops of Krnl to iterate the element in a tensor: the number of loops is determined by `type.getShape().size()`. For scalar, the value of this expression is 0. 
We have to handle this special case in IndexExpr and Krnl.

**Comment by AlexandreEichenberger** [comment]
@chentong319 

> I got error: Expected positive number of original loops.

That error cannot be eliminated. Basically, the program is trying to do a loop structure with 0 or fewer loop nest. That is simply not allowed.

> We build loops of Krnl to iterate the element in a tensor: the number of loops is determined by type.getShape().size(). For scalar, the value of this expression is 0.

Note that no matter the scheme, the number of loops needs to be known at compile time. If that is the case, then you need to isolate the case with 0 loop nests, and if not zero, you can construct an iterate where the number of loops is an arbitrary compile-time constant. Look at the higher level interfaces, such as 

```C++
  // Lambda passes loop indices as 2nd parameter.
  void iterate(ValueRange originalLoops, ValueRange optimizedLoops,
      ValueRange lbs, ValueRange ubs,
      function_ref<void(KrnlBuilder &createKrnl, ValueRange indices)>
          bodyBuilderFn);

  // Lambda passes loop indices as 2nd parameter.
  void iterateIE(ValueRange originalLoops, ValueRange optimizedLoops,
      ArrayRef<IndexExpr> lbs, ArrayRef<IndexExpr> ubs,
      function_ref<void(KrnlBuilder &createKrnl, ValueRange indices)>
          bodyBuilderFn);
```

functions in KrnlHelper.hpp. Both take lists of values or index expressions.

If you want to handle runtime numbers, then there is generally two approaches. First is a switch statement:

```c++
switch (num) {
case 0: code for no loops
              break;
case 1: code for one loop
              break;
...
default:
           exit(1);
}
```

Or one can devise a recursive approach, where you do recursions to handle each level of iterations. Handling memory loads/store also become interesting, as presumably the access functions are also proportional to the dynamic number of loops... If we think hard enough, maybe we can do tail recursion elimination and generate a smart while loop with stacks of iteration indices... Not impossible but harder.

An alternative solution that does not recursion works for loops that are all rectangular.

```C++
num_of_element = product of  extend of the "n" loops (where "n" is only known at runtime)
for(i=0; i<num_of_element; ++i) {
  calculate the indices for each of the "n" indices (for loop dividing i by the sizes of the extend along each dimensions) and putting it in a list.
  perform computations using this list of indices
}
```


I don't think we are ready for this, because with unranked tensors, a lot of things will break at this time.

**Comment by chentong319** [comment]
The semantics of scalar type is simple:  there  is always only one element in the tensor. 
The problem is how we handle scalar type uniformly with "real" tensor. I am thinking of providing a set of interface that treat scalar as one-element tensor, perhaps under the hood of IndexExpr. We just need to be careful to use the original type info for graph-level and hopefully we do not need to change lowering to Krnl that much. We can discuss this issue at group meeting.

**Comment by tungld** [comment]
Regarding loop creation, it seems not related to IndexExpr. We are defining a nested loop with a given number `createKrnl.defineLoops(numberOfLoops)`, so it is clearly that users of krnl need to know `numberOfLoops`, and should check if it it zero or not, except we need to change how to define nested loops. So far all element-wise ops have had a check `if numberOfLoops > 0 then emit krnl loop`. If we call, say Sigmoid, on a scalar tensor, no krnl.iterate is generated.

Only issue is when we have unranked tensors, but so far all tensors come to krnl need to be ranked.

**Comment by chentong319** [comment]
@tungld The error occurred when the loop is initialized with IndexExpr. Of course, we can create the krnl with the special check for scalar. But we did not do such check in many places. It would be better if we find a uniform way to handle scalar and tensor. Will tensor<i32> and tensor<1xi32> be both translated into memref<1xi32>?

**Comment by tungld** [comment]
> It would be better if we find a uniform way to handle scalar and tensor. Will tensor and tensor<1xi32> be both translated into memref<1xi32>?

Is it a good idea to import all ONNX scalars as tensor<1xtype> instead of tensor<type> so that we don't have scalar tensors? Then there is no problem with loops or IndexExpr. Loops with upper bound 1 will be eliminated eventually I think.

**Comment by AlexandreEichenberger** [comment]
> Is it a good idea to import all ONNX scalars as tensor<1xtype> instead of tensor<type> so that we don't have scalar tensors? 

I never really understood the difference between  `tensor<1xtype>` and  `tensor<type>`, and the even stranger definition of `tensor<0xtype>`.

If we do that, let's make sure we have a sufficient number of tests to prevent regressions...

**Comment by chentong319** [comment]
> Is it a good idea to import all ONNX scalars as `tensor<1xtype>` instead of `tensor<type>` so that we don't have scalar tensors? Then there is no problem with loops or IndexExpr. Loops with upper bound 1 will be eliminated eventually I think.
No. Actually it's my first thought and I implemented it. It cause errors in the model. As I mentioned previously:
 `Tense with scalar type is needed because some ops, for example, Clip, require scalar input. Moreover, some shape related ops, such as Squeeze/Unsqueeze, behave differently between tensor and tensor<1xi64>. `

**Comment by tungld** [comment]
Regarding `tensor<0xtype>`, it reminds me of this discussion thread: https://llvm.discourse.group/t/what-is-the-semantics-of-memref-0xf32-and-tensor-0xf32/3557
